Add Move support to the HDFS backend

anjackson opened the pull request:
<!--
Thank you very much for contributing code or documentation to rclone! Please
fill out the following questions to make it easier for us to review your
changes.

You do not need to check all the boxes below all at once, feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box.
-->

#### What is the purpose of this change?

<!--
Describe the changes here
-->

The purpose of this change is to add `Move` support to the HDFS back end, so that e.g. `backup-dir` can be used with HDFS remotes.

#### Was the change discussed in an issue or in the forum before?

<!--
Link issues and relevant forum posts here.
-->

No.

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [ ] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

anjackson commented:
This change has been tested locally against our Hadoop 3 cluster.

This is my first Go project so my apologies if I'm not following any appropriate conventions.

ncw commented:
Looks about right :-)

I'll run the CI on it and it will lint it, then I'll do a review.

ncw commented:
Some quick thoughts.

Can you update the overview here:

https://github.com/rclone/rclone/blob/7366e97dfc80b2ee8c0482d8d5cc7e67b8cde770/docs/content/overview.md?plain=1#L425

And also you should add another interface assertion here for `fs.Mover`

https://github.com/rclone/rclone/blob/7366e97dfc80b2ee8c0482d8d5cc7e67b8cde770/backend/hdfs/fs.go#L315-L316

Is `DirMove` (ie renaming a directory) possible too?

Can you make sure the integration tests run? Best way to do this is to `go install ./...` in the rclone root then

    test_all -backends hdfs

This will run docker stuff, run the backend against it

Looks like the compile failed with `backend/hdfs/fs.go:296:6: no new variables on left side of :=`

anjackson commented:
Gah, sorry for the silly compilation error. I'll take a look at adding those changes.

anjackson commented:
Okay, I've added `DirMove` and tested both locally against our Hadoop 3 cluster.  I've also updated the documentation, and verified that the `test_all -backends hdfs` tests pass.

ivandeex commented:
@ncw 
Is there some deep logic in reviewing a PR submitted few hours ago when there are many PRs waiting for review for 6+ months where authors had already come come through several rounds of addressing previous review comments? Them"Newest submitted" policy  leaves them no chance :-/

ivandeex commented:
@ncw 
Is there some deep logic in reviewing a PR submitted few hours ago when there are many PRs waiting for review for months where authors had already come through several rounds of addressing previous review comments? The "Newest submitted" policy  leaves them no chance :-/

ncw commented on backend/hdfs/fs.go:
```diff
@@ -263,6 +263,95 @@ func (f *Fs) Purge(ctx context.Context, dir string) error {
+		return nil, err
+	}
+
+	// Look up the resulting object
```
The linter is complaining about this bit of code

You need to check the value of err and return it if not nil...

anjackson replied:
Ack, of course. Sorry about that.

ncw reviewed (approved):
This looks great now - thank you :-)

I'll squash and merge now.
